rd-net/RdFramework/Impl/Serializers.cs

```diff
@@ -348,24 +324,75 @@ bool TryGetTypeMapping(Type type1, out RdId rdId)
     }
 
     
-    private readonly HashSet<Type> myRegisteredToplevels = new HashSet<Type>();
     public void RegisterToplevelOnce(Type toplevelType, Action<ISerializers> registerDeclaredTypesSerializers)
     {
-      var r = new ToplevelRegistration(toplevelType, registerDeclaredTypesSerializers);
-#if !NET35
-      var task = myBackgroundRegistrar.SendOrExecuteInline(r);
-      Assertion.Assert(task.IsCompleted, "task.IsCompleted: {0}", task.Status);
-#else
-      RegisterToplevelInternal(r);
- #endif
+      myBackgroundRegistrar.Queue(toplevelType, registerDeclaredTypesSerializers);
+    }
+
+    struct ToplevelRegistration
+    {
+      internal Type Type { get; }
+      internal Action<ISerializers> Action { get; }
+
+      public ToplevelRegistration(Type type, Action<ISerializers> action)
+      {
+        Type = type;
+        Action = action;
+      }
     }
 
-    private void RegisterToplevelInternal(ToplevelRegistration r)
+    private class BackgroundRegistrar
     {
-      if (!myRegisteredToplevels.Add(r.Type)) return;
-      Protocol.InitTrace?.Log($"REGISTER serializers for {r.Type.Name}");
+      [NotNull] private readonly Serializers myOwner;
+      [NotNull] private readonly TaskScheduler myScheduler;
+
+      [NotNull] private readonly object myLock = new object();
+      [NotNull] private readonly Queue<ToplevelRegistration> myQueue = new Queue<ToplevelRegistration>();
+      [NotNull] private readonly HashSet<Type> myRegisteredToplevels = new HashSet<Type>();
+
+      private int myQueueSize;
+
+      public BackgroundRegistrar([NotNull] Serializers owner, [CanBeNull] TaskScheduler scheduler)
+      {
+        myOwner = owner;
+        myScheduler = scheduler ?? TaskScheduler.Default;
+      }
+
+      public void Queue(Type type, Action<ISerializers> action)
+      {
+        lock (myLock)
+        {
+          if (!myRegisteredToplevels.Add(type))
+            return;
 
-      r.Action(this);
+          myQueueSize++;
+          myQueue.Enqueue(new ToplevelRegistration(type, action));
+        }
+
+        var task = new Task(ProcessQueue);
```

Comment:
This is a bit awkward to queue a task each time given the only one will do any job. But given this calls are seldom it it is ok

Reply:
yep, but I prefer simplicity here over tiny performance gain